### PR TITLE
DRep Refunds and update evalTransactionBalance

### LINKED ISSUE #3645: DRep deposits and refunds
Three parts required:

1. Record the deposit amount in the DRep state
2. Validate the deposit and refund in GOVCERT rule
3. `getConsumedValue` and `getProducedValue` will need to account for DReps certificates, in the same way StakingCredential certificates are handled in Shelley.

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Closes #3645 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/UTxO.hs [hunk 1]
```diff
@@ -57,22 +72,59 @@ getConwayScriptsNeeded utxo txb = getAlonzoScriptsNeeded utxo txb <> voterScript
...
+      let go accum@(!drepRegsInTx, !totalDeposit) = \case
+            RegDRepTxCert c _ _
+              -- Re-registration is a no-op and not considered
+              | Just _ <- lookupDRepDeposit c -> accum
+              | otherwise -> (Set.insert c drepRegsInTx, totalDeposit <+> pp ^. ppDRepDepositL)
+            UnRegDRepTxCert c _
+              -- DRep previously registered in the same tx.
+              | Set.member c drepRegsInTx -> (Set.delete c drepRegsInTx, totalDeposit <-> pp ^. ppDRepDepositL)
+            _ -> accum
+       in snd $ foldl' go (Set.empty, Coin 0) $ txb ^. certsTxBodyL
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/UTxO.hs]: This is incorrect. You totally mixed it up. Produced is what leaves the transaction (outputs, fees and deposits). While consumed is what adds to the transaction (inputs, withdrawals and refunds)

In this PR you are doing the opposite.

Implement it by example in the same way the Stake Delegation deposits are implemented here: 

* for produced: https://github.com/input-output-hk/cardano-ledger/blob/d14ff95a51e7c229cc64a07579976d9a2bdd9078/eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/RefundsAndDeposits.hs#L60
* for consumed: https://github.com/input-output-hk/cardano-ledger/blob/d14ff95a51e7c229cc64a07579976d9a2bdd9078/eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/RefundsAndDeposits.hs#L118C1-L118C16

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs [hunk 2]
```diff
@@ -218,6 +220,8 @@ class EraTxBody era => EraUTxO era where
     PParams era ->
     -- | Check whether a pool with a supplied PoolStakeId is already registered.
     (KeyHash 'StakePool (EraCrypto era) -> Bool) ->
+    -- | Function that can lookup current drep deposits
+    (Credential 'DRepRole (EraCrypto era) -> Maybe Coin) ->
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: Produced should not have this argument. This should have been an indicator that something is wrong, cause I've mentioned this before that only one of the two functions needs a new argument.
```suggestion
```

**Review by lehins — CHANGES_REQUESTED** [review]
I realize why you mixed it up. Consumed and produced are a bit weirdly named and unclear concepts.

Anyways, this PR needs some serious fixing.

File: eras/conway/impl/src/Cardano/Ledger/Conway/UTxO.hs (hunk 1 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/UTxO.hs]: Thanks for the pointers! :bowing_man:

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs (hunk 2 above)
Comment (aniketd) [reply libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: Sorry I didn't note that mention, and went down the wrong rabbit-hole on this one.

**Comment by aniketd** [comment]
I hope I have got it correct this time :cold_sweat:

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs [hunk 3]
```diff
@@ -141,7 +219,5 @@ spec =
       prop "evalBalanceTxBody" $ propEvalBalanceTxBody @Alonzo
     describe "BabbageEra" $ do
       prop "evalBalanceTxBody" $ propEvalBalanceTxBody @Babbage
-
--- TODO: bring back when TxCerts become a type family
--- describe "ConwayEra" $ do
---   prop "evalBalanceTxBody" $ propEvalBalanceTxBody @Conway
+    describe "ConwayEra" $ do
+      prop "evalBalanceTxBodyConway" $ propEvalBalanceTxBodyConway @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs]: Get rid of `evalBalanceTxBodyConway` and `propEvalBalanceTxBodyConway`. Copy pasting functionality in order to test that functionality matches the copy is not a test.

All you needed to do was uncomment the piece of code with a TODO

```suggestion
      prop "evalBalanceTxBody" $ propEvalBalanceTxBody @Conway
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/UTxO.hs [hunk 4]
```diff
@@ -163,11 +166,12 @@ shelleyProducedValue pp isRegPoolId txBody =
 getConsumedCoin ::
   ShelleyEraTxBody era =>
   PParams era ->
-  (StakeCredential (EraCrypto era) -> Maybe Coin) ->
+  (Credential 'Staking (EraCrypto era) -> Maybe Coin) ->
+  (Credential 'DRepRole (EraCrypto era) -> Maybe Coin) ->
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/UTxO.hs]: This function is not using the argument, thus it should not be even accepting that argument! This is always wrong to write functions like that. It is at the use site 
 that it should ignore the argument
```suggestion
```

It is in the instance declaration that you should ignore the argument:
```haskell
  getConsumedValue pp lookupRefund _ = getConsumedCoin pp lookupRefund
```

File: eras/mary/impl/src/Cardano/Ledger/Mary/UTxO.hs [hunk 5]
```diff
@@ -54,11 +55,12 @@ instance Crypto c => EraUTxO (MaryEra c) where
 getConsumedMaryValue ::
   (MaryEraTxBody era, Value era ~ MaryValue (EraCrypto era)) =>
   PParams era ->
-  (StakeCredential (EraCrypto era) -> Maybe Coin) ->
+  (Credential 'Staking (EraCrypto era) -> Maybe Coin) ->
+  (Credential 'DRepRole (EraCrypto era) -> Maybe Coin) ->
```
Comment (lehins) [inline eras/mary/impl/src/Cardano/Ledger/Mary/UTxO.hs]: Same comment applies here.
```suggestion
```

**Review by lehins — CHANGES_REQUESTED** [review]
Now that looks much better! :smile: 

Few minor things need adjustment. Also ghc-9.6 needs TyepOperators extension, see the CI failure

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/UTxO.hs (hunk 4 above)
Comment (aniketd) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/UTxO.hs]: Got it! :raised_hands:

Comment (aniketd) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/UTxO.hs]: This is actually even worse. I had `getConsumedMaryValue` being used inside `conwayConsumedValue` and didn't realise the lookup function need not be provided there!

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
```
Comment (aniketd) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs]: Removing this line.

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs [hunk 7]
```diff
@@ -31,6 +32,7 @@ import qualified Data.Set as Set
 import Lens.Micro
 import Test.Cardano.Ledger.Babbage.Arbitrary ()
 import Test.Cardano.Ledger.Common
+import Test.Cardano.Ledger.Conway.Arbitrary ()
```
Comment (aniketd) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs]: Removing this line too.

**Review by lehins — APPROVED** [review]
Looks great. Thank you!

**Pull request merged**